### DISCUSSION TITLE
Add an optional data field for `SourceItem`

### DISCUSSION THREAD

**Discussion by xuansontrinh** [build-server-protocol/build-server-protocol#691] [Proposals]
Just like the optional data field in `BuildTarget`, it would be advantageous for `SourceItem` to include a similar field. This new field could hold specific details about the source, such as JVM-related information about the source file like package prefix.

Why is this useful? This field would allow us to supply additional information to IDEs, enabling them to offer more intelligent features on source file level.

**Comment by adpi2** [comment]
In theory I think any response could have its generic data field that a language extension could use to transfer specific information.

For your specific use case, is it really possible to define a package prefix in Maven or Gradle?

Do you have any other example of custom fields we could need in `SourceItem`?

**Reply by xuansontrinh** [reply]
My scenario is with Bazel build tool.

Unlike Maven or Gradle, which tie package names to directory structures, Bazel allows package names to be defined independently of file locations. This means a file's package in Bazel might not align with its directory path, making it difficult to infer package names automatically.

For example:

```md
root
- arbitrary1
  - file1.java (package my_project.x)
  - file2.java (package my_project.y)
- arbitrary2
  - arbitrary3
    - file3.java (package my_project.z)
```

Here, `file1.java`, `file2.java` and `file3.java` are in different directories but belong to different packages. This flexibility is common in Bazel but would be unusual in Maven or Gradle. And because of such flexibility, the info from package prefixes of source files should be passed to IDE to support proper highlighting and smart features, in my case the IDE is IntelliJ IDEA.

> Do you have any other example of custom fields we could need in `SourceItem`?

Custom fields in `SourceItem` might include external annotations or documentation links that are relevant to the source files but not part of the code itself.

**Reply by adpi2** [reply]
Yes I think it is a good use case.

**Reply by xuansontrinh** [reply]
Then shall I go ahead and make a PR to add this field?
Or are there any concerns that I need to address before proceeding?

**Reply by adpi2** [reply]
You can go ahead and make a PR.
